tritium-cache/src/main/java/com/palantir/tritium/cache/AsyncCacheImpl.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.cache;
+
+import static com.palantir.logsafe.Preconditions.checkState;
+
+import com.google.common.collect.Iterators;
+import com.google.errorprone.annotations.MustBeClosed;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import org.jspecify.annotations.Nullable;
+
+class AsyncCacheImpl<K, V> implements AsyncCache<K, V> {
+
+    private final com.github.benmanes.caffeine.cache.AsyncCache<K, V> cache;
+
+    AsyncCacheImpl(com.github.benmanes.caffeine.cache.AsyncCache<K, V> cache) {
+        this.cache = cache;
+    }
+
+    @Override
+    @Nullable
+    public final V getIfPresent(K key) {
+        return cache.synchronous().getIfPresent(key);
+    }
+
+    @Override
+    @Nullable
+    public final V get(K key, Function<? super K, ? extends V> mappingFunction) {
+        CompletableFuture<V> future;
+        try (ValueFactory<K, V> valueFactory = new ValueFactory<>(mappingFunction)) {
+            future = cache.get(key, valueFactory);
+        }
+
+        return await(future);
+    }
```

Comment:
This is how we get the desired `invalidateAll` behavior for async caches. We don't actually start to load the value until the future has been inserted into the cache.

Reply:
fwiw, the reason Caffeine couldn't do it this way is because the future is supplied by the user in their `AsyncCacheLoader`. Since `CompletableFuture` is identity equality it is important to return the user's future instance back to them. For instance if they want to cancel or remove the instance by the `asMap` view.